src/ts_catalog/compression_settings.c

```diff
@@ -0,0 +1,255 @@
+														  TupleDesc desc);
+
+CompressionSettings *
+ts_compression_settings_create(Oid relid, ArrayType *segmentby, ArrayType *orderby,
+							   ArrayType *orderby_desc, ArrayType *orderby_nullsfirst)
+{
+	Catalog *catalog = ts_catalog_get();
+	CatalogSecurityContext sec_ctx;
+	Relation rel;
+	FormData_compression_settings fd;
+
+	Assert(OidIsValid(relid));
+
+	/*
+	 * The default compression settings will always have orderby settings but the user may have
+	 * chosen to overwrite it. For both cases all 3 orderby arrays must either have the same number
+	 * of entries or be all NULL.
+	 */
+	Assert((orderby && orderby_desc && orderby_nullsfirst) ||
+		   (!orderby && !orderby_desc && !orderby_nullsfirst));
+
+	fd.relid = relid;
+	fd.segmentby = segmentby;
+	fd.orderby = orderby;
+	fd.orderby_desc = orderby_desc;
+	fd.orderby_nullsfirst = orderby_nullsfirst;
+
+	rel = table_open(catalog_get_table_id(catalog, COMPRESSION_SETTINGS), RowExclusiveLock);
+
+	HeapTuple new_tuple = compression_settings_formdata_make_tuple(&fd, RelationGetDescr(rel));
+	ts_catalog_database_info_become_owner(ts_catalog_database_info_get(), &sec_ctx);
+	ts_catalog_insert(rel, new_tuple);
+	ts_catalog_restore_user(&sec_ctx);
+	heap_freetuple(new_tuple);
+
+	table_close(rel, RowExclusiveLock);
+
+	return ts_compression_settings_get(relid);
+}
+
+static void
+compression_settings_fill_from_tuple(CompressionSettings *cs, TupleInfo *ti)
+{
+	FormData_compression_settings *fd = &cs->fd;
+	Datum values[Natts_compression_settings];
+	bool nulls[Natts_compression_settings];
+	bool should_free;
+
+	HeapTuple tuple = ts_scanner_fetch_heap_tuple(ti, false, &should_free);
+
+	heap_deform_tuple(tuple, ts_scanner_get_tupledesc(ti), values, nulls);
+
+	MemoryContext old = MemoryContextSwitchTo(ti->mctx);
+
+	fd->relid = DatumGetInt32(values[AttrNumberGetAttrOffset(Anum_compression_settings_relid)]);
+
+	if (nulls[AttrNumberGetAttrOffset(Anum_compression_settings_segmentby)])
+		fd->segmentby = NULL;
+	else
+		fd->segmentby = DatumGetArrayTypeP(
+			values[AttrNumberGetAttrOffset(Anum_compression_settings_segmentby)]);
+
+	if (nulls[AttrNumberGetAttrOffset(Anum_compression_settings_orderby)])
+		fd->orderby = NULL;
+	else
+		fd->orderby =
+			DatumGetArrayTypeP(values[AttrNumberGetAttrOffset(Anum_compression_settings_orderby)]);
+
+	if (nulls[AttrNumberGetAttrOffset(Anum_compression_settings_orderby_desc)])
+		fd->orderby_desc = NULL;
+	else
+		fd->orderby_desc = DatumGetArrayTypeP(
+			values[AttrNumberGetAttrOffset(Anum_compression_settings_orderby_desc)]);
+
+	if (nulls[AttrNumberGetAttrOffset(Anum_compression_settings_orderby_nullsfirst)])
+		fd->orderby_nullsfirst = NULL;
+	else
+		fd->orderby_nullsfirst = DatumGetArrayTypeP(
+			values[AttrNumberGetAttrOffset(Anum_compression_settings_orderby_nullsfirst)]);
+
+	MemoryContextSwitchTo(old);
+
+	if (should_free)
+		heap_freetuple(tuple);
+}
+
+TSDLLEXPORT CompressionSettings *
+ts_compression_settings_get(Oid relid)
```

Comment:
It would be nice if we had an option to fill in the compression settings into pre-allocated or static memory. Currently, the compression settings are a number non-contiguous pallocs in CurrentMemoryContext mixed in with temporary memory allocations (e.g., used for scans). so there's no easy way to embed this in an existing struct or cache it. I am currently playing around with caching compression settings in a relcache entry's `rd_amcache` pointer. But that must be a single chunk of memory in CacheMemoryContext.

In the previous approach, at least every form was aligned memory so it was easier to copy.